npm/vite-dev-server/src/resolveConfig.ts

```diff
@@ -90,7 +90,15 @@ export const createViteDevServerConfig = async (config: ViteDevServerConfig, vit
     ].filter((p) => p != null),
   }
 
-  const finalConfig = vite.mergeConfig(viteBaseConfig, viteOverrides as Record<string, any>)
+  let resolvedOverrides: UserConfig = {}
+
+  if (typeof viteOverrides === 'function') {
+    resolvedOverrides = await viteOverrides(viteBaseConfig)
```

Comment:
Would it make sense to utilize the [loadConfigFromFile](https://github.com/vitejs/vite/blob/8f315a20a7c3a33c72e7993860b6b69f5d6f0e05/packages/vite/src/node/config.ts#L872) function before executing the viteConfig function? That way, the config passed to the function would have the resolved values from the `configFile` that they could modify.

Reply:
Cool idea, but the `loadConfigFromFile` function is new in Vite 3; since we support Vite 2, we won't be able to use it if we want to have the same experience across both versions.